_posts/2025-08-04-Silly-Computations.md

```diff
@@ -0,0 +1,141 @@
+
+## Database Costs
+
+Let’s say a request comes in to our backend at `/organizations/:orgId`. 
+For a Yesod-backed application like ours, the processing of the request looks something like this:
+- First, run any pre-request middleware actions.
+  - In these, authorization checks are common and we'll want to ensure the current user has access to the organization specified in the URI. 
+- Then we'll run the handler action which contains all of our business logic for that route.
+- Finally, we'll run any post-request middleware actions and send the response to the client.
+
+Now, given the specific organization ID, multiple portions of this request handling might want access to the full `Entity Organization`. Every time we fetch those details we: 
+- `select * from organizations where id = ?`
+- Serialize the response object on our database
+- Transmit it over the network.
+- Deserialize it on our backend 
+- Parse it into a Haskell data type.
+
+Each of these blocks takes around 1-3ms every time it happens. In a well-built application, we might do this 3-5 times per request given the flow I described above. Specifically with Yesod, the various contexts are disconnected from each other. Requests that you made in your authorization checks are disconnected from the ones you make in your handler. Then, within the application logic there are infinite opportunities for silly work, especially when multiple engineers are operating on the same codebase every day. 
+
+## Two Solutions
+
+There are two broad ways I've found to mitigate these silly computations.
+
+### 1. Per-request Caching
+
+We use Yesod’s built-in per-request cache. 
+
+```haskell
+getCachedOrganization :: OrganizationId -> Handler (Maybe Organization)
+getCachedOrganization orgId = do
+  let bytes = UUID.toASCIIBytes $ unOrganizationKey orgId
+  cachedBy bytes (runDB . get $ orgId)
+``` 
+
+(`cachedBy`'s particular implementation is less interesting though it's fairly straightforward to build on top of Yesod's specific cache or to make your own within your `Handler` monad.)
+
+This allows us to share the results of database lookups across the lifecycle of a single request. Without it, there's really no way (within Yesod's framework) to share data between your middleware and your handler contexts. 
+
+### 2. Just Pass the Whole Thing
+
+The simpler fix, and the one I wish we’d use more, is: **stop throwing away the result.**
+
+If you fetch the `Organization` early, just pass the whole entity through the stack. Even if some functions only use the ID, or ignore it entirely.
+
+This might feel wasteful. Isn’t it inefficient to keep passing data we don’t use?
+
+Not really. Haskell is lazy. If no one touches the data, it won’t be evaluated. And compared to the costs of a database round-trip, the overhead is tiny.
```

Comment:
Is this actually true? Wouldn't we have to fully evaluate the `Organization` when we deserialize from the database?

Reply:
Totally agree with that. For some reason, I read this as you saying

> if we don't use the fields, they won't get evaluated